internal/pkg/consul/client.go

```diff
@@ -232,20 +316,52 @@ func (client *consulClient) GetConfigurationValue(name string) ([]byte, error) {
 	return keyPair.Value, nil
 }
 
-// Puts a specific configuration value into Consul
+// PutConfigurationValue puts a specific configuration value into Consul
 func (client *consulClient) PutConfigurationValue(name string, value []byte) error {
 	keyPair := &consulapi.KVPair{
 		Key:   client.fullPath(name),
 		Value: value,
 	}
 
 	_, err := client.consulClient.KV().Put(keyPair, nil)
+
+	retry, err := client.reloadAccessTokenOnAuthError(err)
+	if retry {
+		// Try again with new Access Token
+		_, err = client.consulClient.KV().Put(keyPair, nil)
+	}
+
 	if err != nil {
 		return fmt.Errorf("unable to put value for %s into Consul: %v", client.fullPath(name), err)
 	}
+
 	return nil
 }
 
+func (client *consulClient) reloadAccessTokenOnAuthError(err error) (bool, error) {
+	if err == nil {
+		return false, nil
+	}
+
+	if strings.Contains(err.Error(), aclError) && client.getAccessToken != nil {
+		client.consulConfig.Token, err = client.getAccessToken()
+		if err != nil {
+			err = fmt.Errorf("failed to renew access token: %s", err.Error())
+			return false, err
+		}
```

Comment:
so when in `err != nil` case, the value of `client.consulConfig.Token` seems to become unpredictable depending on how `client.getAccessToken()` is handled.  Should we assign the right value into `client.consulConfig.Token` only when `err==nil`?  This becomes more critical especially this func is called inside go routine and err returns as part of channel inside that go routine.

Reply:
> I would have expected that the we'd actually compare the old and new access tokens in order to determine whether or not to retry.

@bnevis-i , the new token will always be different since it is calling Vault for a new Consul access token.

Reply:
> Should we assign the right value into `client.consulConfig.Token` only when `err==nil`?

@jim-wang-intel , yes I like that. Will make the change